server/sonar-server/src/main/java/org/sonar/server/rule/DefaultRuleFinder.java

```diff
@@ -103,61 +125,97 @@ public DefaultRuleFinder(IndexClient indexes, UserSession userSession) {
+    String description = rule.getDescription();
+    RuleDto.Format descriptionFormat = rule.getDescriptionFormat();
+
     org.sonar.api.rules.Rule apiRule = new org.sonar.api.rules.Rule();
     apiRule
-      .setName(rule.name())
-      .setLanguage(rule.language())
-      .setKey(rule.key().rule())
-      .setConfigKey(rule.internalKey())
+      .setName(rule.getName())
+      .setLanguage(rule.getLanguage())
+      .setKey(rule.getRuleKey())
+      .setConfigKey(rule.getConfigKey())
       .setIsTemplate(rule.isTemplate())
-      .setCreatedAt(rule.createdAt())
-      .setUpdatedAt(rule.updatedAt())
-      .setDescription(rule.htmlDescription())
-      .setRepositoryKey(rule.key().repository())
-      .setSeverity(rule.severity() != null ? RulePriority.valueOf(rule.severity()) : null)
-      .setStatus(rule.status().name())
-      .setTags(rule.tags().toArray(new String[rule.tags().size()]))
-      .setId(((RuleDoc) rule).id());
+      .setCreatedAt(new Date(rule.getCreatedAt()))
+      .setUpdatedAt(new Date(rule.getUpdatedAt()))
+      .setRepositoryKey(rule.getRepositoryKey())
+      .setSeverity(severity != null ? RulePriority.valueOf(severity) : null)
+      .setStatus(rule.getStatus().name())
+      .setTags(rule.getTags().toArray(new String[rule.getTags().size()]))
+      .setId((rule.getId()));
+    if (description != null && descriptionFormat != null) {
+      if (RuleDto.Format.HTML.equals(descriptionFormat)) {
+        apiRule.setDescription(description);
+      } else {
+        apiRule.setDescription(Markdown.convertToHtml(description));
+      }
+    }
 
     List<org.sonar.api.rules.RuleParam> apiParams = newArrayList();
-    for (RuleParam param : rule.params()) {
-      apiParams.add(new org.sonar.api.rules.RuleParam(apiRule, param.key(), param.description(), param.type().type())
-        .setDefaultValue(param.defaultValue()));
+    for (RuleParamDto param : params) {
+      apiParams.add(new org.sonar.api.rules.RuleParam(apiRule, param.getName(), param.getDescription(), param.getType())
+        .setDefaultValue(param.getDefaultValue()));
     }
     apiRule.setParams(apiParams);
 
     return apiRule;
   }
 
-  private RuleQuery toQuery(org.sonar.api.rules.RuleQuery apiQuery) {
-    RuleQuery query = new RuleQuery();
-    if (apiQuery.getConfigKey() != null) {
-      query.setInternalKey(apiQuery.getConfigKey());
-    }
-    if (apiQuery.getKey() != null) {
-      query.setRuleKey(apiQuery.getKey());
+  private enum RuleDtoToKey implements Function<RuleDto, RuleKey> {
+    INSTANCE;
+
+    @Override
+    public RuleKey apply(@Nonnull RuleDto input) {
+      return input.getKey();
     }
-    if (apiQuery.getRepositoryKey() != null) {
-      query.setRepositories(ImmutableList.of(apiQuery.getRepositoryKey()));
+  }
+
+  private enum RuleParamDtoToRuleId implements Function<RuleParamDto, Integer> {
+    INSTANCE;
+
+    @Override
+    public Integer apply(@Nonnull RuleParamDto input) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The "input" parameter nullability is different in the superclass method, and that should not be changed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)